lib/posix/sched.c

```diff
@@ -41,3 +43,47 @@ int sched_get_priority_max(int policy)
 	errno = EINVAL;
 	return -1;
 }
+
+/**
+ * @brief Get scheduling parameters
+ *
+ * See IEEE 1003.1
+ */
+int sched_getparam(pid_t pid, struct sched_param *param)
+{
+	struct sched_param dummy_param = {0};
+	pthread_t tid = 0;
+	int t_policy = -1;
+	int ret = -1;
+
+	if (param == NULL) {
+		param = &dummy_param;
+	}
+
+	tid = (pid == 0) ? pthread_self() : z_get_tid(pid);
+
+	ret = pthread_getschedparam(tid, &t_policy, param);
+	errno = ret;
+
+	return ((ret) ? -1 : 0);
```

Comment:
It might be more in line with the POSIX specification to simply return `-1` and set errno to `ENOSYS` here, since Zephyr does not (yet) support processes or process scheduling.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/sched_getparam.html
```suggestion
	ARG_UNUSED(pid);
	ARG_UNUSED(param);

	errno = ENOSYS;

	return -1;
```

Reply:
> It might be more in line with the POSIX specification to simply return `-1` and set errno to `ENOSYS` here, since Zephyr does not (yet) support processes or process scheduling.

what do you think if we deal only with `pid==0` for `pthread_self()` and errno `ENVAL` otherwise ?

Reply:
> > It might be more in line with the POSIX specification to simply return `-1` and set errno to `ENOSYS` here, since Zephyr does not (yet) support processes or process scheduling.
> 
> what do you think if we deal only with `pid==0` for `pthread_self()` and errno `EINVAL` otherwise ?

There is still a chance for a great deal of inconsistency, even from the perspective of just querying the policy / priority.

For example, Thread A and Thread B are both threads that belong to pid zero (i.e. are part of the same process), but when one looks at them, they see that the individual thread-level scheduling policy and priority are different. So inherently, we are conflating process scheduling and thread scheduling.

The situation can even get a bit more worrisome when one considers that some threads might be running in Kernel mode, and others in user mode.

I think in order to do what Linux does (and I'm not saying it's wrong - that would be kind of nice), there would need to be support from the kernel for essentially threads being equivalent to lightweight processes (which is what Linux does), and for the kernel to support processes as well.

It's a fantastically interesting area for potential improvement, but I would not want to jump ahead of the kernel / architecture on it.

Reply:
ok, removed.

> The situation can even get a bit more worrisome when one considers that some threads might be running in Kernel mode, and others in user mode.

yeah, its another design question on what to do with kernel/user threads at the posix subsys level.